File: fastExp.cpp
// fast exp
#include <iostream>
#include <iterator>
#include <vector>
#include <map>
#include <cmath>
using namespace std;

unsigned long int mfer(unsigned long int number, unsigned long int power, unsigned long int quotient);

unsigned long int mfer(unsigned long int number, unsigned long int power, unsigned long int quotient)
{
    cout << "Called" << endl;

    map<unsigned long int, unsigned long int> memo;
    memo.insert({1, number});
    unsigned long int start_with = 1;
    unsigned long int new_insert = 1;
    unsigned long int remainder = 1;
    int bin_rep[100], bin_len, i;

    cout << "Decalrations done" << endl;

    while (start_with <= power)
    {
        cout << "start_with : " << start_with << endl;
        new_insert = memo.at(start_with);
        new_insert = (new_insert * new_insert) % quotient;
        start_with = start_with * 2;
        // mfer, you forgot to insert
        memo.insert({start_with, new_insert});
    }

    cout << "MemTable calculated" << endl;

    bin_len, i = 0;
    while (power > 0)
    {
        bin_rep[i] = power % 2;
        power = power / 2;
        i++;
        bin_len++;
    }

    cout << "Power binary rep done" << endl;

    for (i = bin_len; i >= 0; i--)
    {
        if (bin_rep[i] == 1)
        {
            remainder = remainder * memo.at(pow(2, i));
        }
    }
    return (remainder % quotient);
}

int main()
{
    cout << mfer(24, 34, 215) << endl;
}

File: fastExp.py
# fast exp
import time

def mfer(number, power, quotient):
    remainder = 1
    power_dict = {1: number%quotient}
    start_with = 1

    while start_with <= power:
        # print(power_dict[sorted(list(power_dict.keys()))[-1]])
        new_insert = (power_dict[sorted(list(power_dict.keys()))[-1]] ** 2) % quotient
        start_with *= 2
        print(start_with, new_insert)
        power_dict[start_with] = new_insert

    bin_rep = bin(power)[2:][::-1]
    for x in range(len(bin_rep)):
        if bin_rep[x] == '1':
            print(2**x)
            remainder *= power_dict[2**x]

    return remainder % quotient

if __name__ == "__main__":
    x0 = time.time()
    print(mfer(24, 34, 215))
    print(time.time() - x0)